collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt

```diff
@@ -0,0 +1,220 @@
+            var formId: String? = null
+            var formName: String? = null
+            var version: String? = null
+            var downloadUrl: String? = null
+            var manifestUrl: String? = null
+            var hash: String? = null
+            // don't process descriptionUrl
+
+            val fieldCount = xformElement.childCount
+            for (j in 0 until fieldCount) {
+                if (xformElement.getType(j) != Element.ELEMENT) {
+                    // whitespace
+                    continue
+                }
+
+                val child = xformElement.getElement(j)
+                if (!isXformsListNamespacedElement(child)) {
+                    // someone else's extension?
+                    continue
+                }
+
+                when (child.name) {
+                    "formID" -> {
+                        formId = XFormParser.getXMLText(child, true)
+                        if (formId != null && formId.length == 0) {
+                            formId = null
+                        }
+                    }
+                    "name" -> {
+                        formName = XFormParser.getXMLText(child, true)
+                        if (formName != null && formName.length == 0) {
+                            formName = null
+                        }
+                    }
+                    "version" -> {
+                        version = XFormParser.getXMLText(child, true)
+                        if (version != null && isBlank(version)) {
+                            version = null
+                        }
+                    }
+                    "majorMinorVersion" -> {
+                    }
+                    "descriptionText" -> {
+                    }
+                    "downloadUrl" -> {
+                        downloadUrl = XFormParser.getXMLText(child, true)
+                        if (downloadUrl != null && downloadUrl.length == 0) {
+                            downloadUrl = null
+                        }
+                    }
+                    "manifestUrl" -> {
+                        manifestUrl = XFormParser.getXMLText(child, true)
+                        if (manifestUrl != null && manifestUrl.length == 0) {
+                            manifestUrl = null
+                        }
+                    }
+                    "hash" -> {
+                        hash = XFormParser.getXMLText(child, true)
+                        if (hash != null && (hash.length == 0 || !hash.startsWith("md5:"))) {
+                            hash = null
+                        } else {
+                            hash = hash.substring("md5:".length)
+                        }
+                    }
+                }
+            }
+
+            if (formId == null || downloadUrl == null || formName == null) {
+                formList.clear()
+                return null
+            }
+
+            val formListItem =
```

Comment:
Does that make sense to construct this object ad all? You can just add it directly to the list:
`            formList.add(FormListItem(downloadUrl, formId, version!!, hash, formName, manifestUrl))
`